### ISSUE TITLE
os.path functions get wrong type assignments.

### ISSUE THREAD

**Issue by Michallote** [astral-sh/ty#405]
### Summary

Minimum code fail:

```python
#fail.py
import os

def get_rel(output_directory: str) -> str:
    assert os.path.exists(
        output_directory
    ), f"Output directory '{output_directory}' does not exist. Create the folder and try again."
    output_directory = os.path.relpath(output_directory)
    return output_directory
```

And running in the terminal:

```bash
$ uvx ty check fail.py
WARN ty is pre-release software and not ready for production use. Expect to encounter bugs, missing features, and fatal errors.
error[invalid-assignment]: Object of type `bytes` is not assignable to `str`
  --> fail.py:9:5
   |
 7 |         output_directory
 8 |     ), f"Output directory '{output_directory}' does not exist. Create the folder and try again."
 9 |     output_directory = os.path.relpath(output_directory)
   |     ^^^^^^^^^^^^^^^^
10 |     return output_directory
   |
info: rule `invalid-assignment` is enabled by default

Found 1 diagnostic
```

commenting os.path.exists does not change the outcome.

### Version

ty 0.0.1-alpha.2

**Comment by MichaReiser** [comment]
Playground: https://play.ty.dev/79aa2c89-ec65-438f-91e6-81d9a6341275

**Comment by sharkdp** [comment]
Thank you for reporting this. `os.path.relpath` is overloaded in typeshed:
```pyi
@overload
def relpath(path: LiteralString, start: LiteralString | None = None) -> LiteralString: ...
@overload
def relpath(path: BytesPath, start: BytesPath | None = None) -> bytes: ...
@overload
def relpath(path: StrPath, start: StrPath | None = None) -> str: ...
```

We mistakenly use the second overload here which returns `bytes`. The reason we do this is that we don't understand the `BytesPath` type alias yet (since it's a `typing.TypeAlias` alias, not a PEP 695 `type` alias):

```pyi
BytesPath: TypeAlias = bytes | PathLike[bytes]
```

see https://github.com/astral-sh/ty/issues/221 which keeps track of our type alias support.

**Comment by Michallote** [comment]
But the operation overloading should return StrPath if my input is typed as a string no? Or the BytesPath makes an union with str too therefore defaulting to it first as it doesn't check the last one?

**Comment by jelle-openai** [comment]
Because ty doesn't understand `TypeAlias`, it treats both `BytesPath` and `StrPath` as essentially Any and picks the first one that matches.

**Comment by AlexWaygood** [comment]
Thanks again for the report! I'm closing this in favour of https://github.com/astral-sh/ty/issues/544, since the fix for this will just be to implement #544

**Comment by carljm** [comment]
Confirmed that this is fixed by https://github.com/astral-sh/ruff/pull/21394
